synapse/storage/end_to_end_keys.py

```diff
@@ -281,3 +285,145 @@ def delete_e2e_keys_by_device_txn(txn):
         return self.runInteraction(
             "delete_e2e_keys_by_device", delete_e2e_keys_by_device_txn
         )
+
+    def _set_e2e_device_signing_key_txn(self, txn, user_id, key_type, key):
+        """Set a user's cross-signing key.
+
+        Args:
+            txn (twisted.enterprise.adbapi.Connection): db connection
+            user_id (str): the user to set the signing key for
+            key_type (str): the type of key that is being set: either 'master'
+                for a master key, 'self_signing' for a self-signing key, or
+                'user_signing' for a user-signing key
+            key (dict): the key data
+        """
+        # the cross-signing keys need to occupy the same namespace as devices,
+        # since signatures are identified by device ID.  So add an entry to the
+        # device table to make sure that we don't have a collision with device
+        # IDs
+        for v in key["keys"].values():
```

Comment:
I think I would spell this as `pubkey = next(iter(key["keys"].values()))`, but ymmv.

More helpfully: could you add a comment documenting the shape of `key` to show why this is the right thing to do?